proposals/POEM-2-function-pulling-container-scheduler.md

```diff
@@ -0,0 +1,132 @@
+* [Component Design](https://cwiki.apache.org/confluence/display/OPENWHISK/Component+Design)
+
+### Implementation details
+
+For the record, we(Naver) are already operating OpenWhisk with this scheduler in a production environment.
+We want to contribute the scheduler and hope it evolves with the community.
+
+There are many [new components](https://cwiki.apache.org/confluence/display/OPENWHISK/Component+Design).
+
+#### Common components
+We store data in ETCD, there are many relevant components such as `EtcdClient`, `LeaseKeepAliveService`, `WatcherService`, `DataManagementService`, etc.
+
+* `EtcdClient`: An ETCD client offering basic CRUD operations.
+* `DataManagementService`: In charge of storing/recovering data in ETCD. It internally utilizes `LeaseKeepAliveService`, and `WatcherService`.
+* `LeaseKeepAliveService`: In charge of keeping the given lease alive.
+* `WatcherService`: Watches the given keys and receives events for the keys.
+
+
+#### Scheduler components
+
+In an abstract view, schedulers provide queueing, container scheduling, and activation routing.
+
+* `QueueManager`: The main entry point for queue creation request. It has references to all queues.
+* `MemoryQueue`: Dynamically created/deleted for each action. It watches the incoming/outgoing requests and triggers container creation.
+* `ContainerManager`: Schedule container creation requests to appropriate invokers.
+* `ActivationServiceImpl`: Provide API for containers to fetch activations via Akka-grpc. It works in a long-poll way to avoid busy-waiting.
+
+#### Controller components
+* `FPCPoolBalancer`: Create queues if not exist, and forward messages to them.
+* `FPCEntitlementProvider`: Throttle activations based on throttling information in ETCD.
+
+#### Invoker components.
+
+* `FunctionPullingContainerPool`: A container pool for function pulling container. It handles the container creation requests.
+* `FunctionPullingContainerProxy`: A proxy for a container. It repeatedly fetches activations and invokes them.
+* `ActivationClientProxy`: The Akka-grpc client. It communicates with `ActivationServiceImpl` in schedulers.
+* `InvokerHealthManager`: Manages the health and resources information of invokers. The data is stored in ETCD. If an invoker becomes unhealthy, it invokes health activations.
+
+## Future work.
+
+#### Persistence
+
+We reached a conclusion that the persistence of activation requests is not that mandatory requirement along with the at-most-once nature and circuit-breaking of OpenWhisk.
+But if it is desired, we can implement a persistent queue rather than an in-memory queue.
+
+#### Multiple partitions
+
+Currently, a queue only has one partition. Since there can be multiple queues for each action, cluster-wide RPS(request per second) linearly increases.
+If high RPS for one action is required, we can implement partitioning in queues to introduce parallel processing like what Kafka does.
```

Comment:
I don't understand the reference to Kafka here.

Reply:
Partitions are the unit of parallelism in Kafka.
So if we want multiple consumers running, we need at least the same number of partitions.

In our current implementation, each queue has only one partition(no partition). So the performance can be limited by the performance of one queue. Generally, it would be enough for most of cases, but as the scale of cluster grows, multiple partitions may be desired in the huge scale.

Reply:
Is there any idea what the new queues will look like? Seems like overkill to be implementing partitioned queues when there's existing queues out there like kafka that will do this. I get that you just want in mem though so using kafka would be overkill

Reply:
The partitioning is future work. This feature would not be included in my proposal at this time.
And I think the version in this proposal would be enough for most cases.

If you are mentioning about the new queue which is introduced by my proposal, you may find this useful: [2.5 Rule out Kafka](https://cwiki.apache.org/confluence/display/OPENWHISK/Design+consideration).
In short, I introduced a new queue for better controllability, and routing.